File: python/gol.py
from time import sleep

width = 30
height = 25
states = [' ', '#']

def get_neighbors(board: list[list[str]], y: int, x: int) -> int:
    count: int  = 0
    width: int  = len(board[0])
    height: int = len(board)

    for i in range(y-1, y+2):
        for j in range(x-1, x+2):
            if board[(i + height )% height][(j + width) % width] == states[1] and not (i == y and j == x):
                count += 1
    return count

def print_board(board: list[list[str]]) -> None:
    for row in board:
        for col in row:
            print(col,end=' ') 
        print()

def clear_screen() -> None:
    print("\033[1;1H\033[2J")


def step(board_a: list[list[str]], board_b: list[list[str]]) -> None:
    for i in range(height):
        for j in range(width):
            neighbors = get_neighbors(board_a,i,j)
            if board_a[i][j] == states[1] and (neighbors == 2 or 3 ==
                neighbors):
                board_b[i][j] = states[1]
            elif neighbors == 3:
                board_b[i][j] = states[1]
            else:
                board_b[i][j] = states[0]

if __name__ == "__main__":
    board_a: list[list[str]] = [[states[0] for i in range(width)] for j in
        range(height)]
    board_b: list[list[str]] = [[states[0] for i in range(width)] for j in
        range(height)]

    # R-Pentomino
    board_a[(height//2) + 0][(width // 2) + 1] = states[1]
    board_a[(height//2) + 0][(width // 2) + 0] = states[1]
    board_a[(height//2) + 1][(width // 2) + 0] = states[1]
    board_a[(height//2) + 2][(width // 2) + 0] = states[1]
    board_a[(height//2) + 1][(width // 2) - 1] = states[1]

    while(True):
        step(board_a, board_b)
        board_a, board_b = board_b, board_a

        clear_screen()
        print_board(board_a)

        sleep(0.1)

File: js/gol.js

var states = [' ', '#'];
var width = 30;
var height = 20;


const sleep = ms => new Promise(r => setTimeout(r, ms));

function print_board(board) {
    board.forEach(row => {
        row.forEach(s => {
            process.stdout.write(' ' + s)
        });
        console.log();
    });
}

function neighbors(board, x, y) {
    count = 0;
    for (let i = x - 1; i < x + 2; i++) {
        for (let j = y - 1; j < y + 2; j++) {
            if (j == y && i == x) {
                continue;
            }
            if (board[(i + board.length) % board.length][(j + board[0].length) % board[0].length] == states[1]) {
                count++;
            }
        }
    }
    return count;
}

function clear_screen() {
    console.log("\x1b[1;1H\x1b[2J");
}

function step(board_a, board_b) {
    for (let i = 0; i < board_a.length; i++) {
        for (let j = 0; j < board_a[0].length; j++) {
            n = neighbors(board_a, i, j);

            if (board_a[i][j] == states[1] && (n == 2 || 3 ==
                n)) {
                board_b[i][j] = states[1];
            }
            else if (n == 3) {
                board_b[i][j] = states[1];
            }
            else {
                board_b[i][j] = states[0];
            }

        }
    }
}

function create_boards(width, height) {
    board_a = Array(height);
    board_b = Array(height).fill(Array(width).fill(states[0]));
    for (let i = 0; i < height; i++) {
        board_a[i] = Array(width);
        board_b[i] = Array(width);
        for (let j = 0; j < width; j++) {
            board_a[i][j] = states[0];
            board_b[i][j] = states[0];
        }
    }
    return [board_a, board_b]
}

function add_r_pentomino(board) {
    height = board.length
    width = board[0].length
    board[(height / 2) + 0][(width / 2) + 1] = states[1];
    board[(height / 2) + 0][(width / 2) + 0] = states[1];
    board[(height / 2) + 1][(width / 2) + 0] = states[1];
    board[(height / 2) + 2][(width / 2) + 0] = states[1];
    board[(height / 2) + 1][(width / 2) - 1] = states[1];

}

async function main() {
    [board_a, board_b] = create_boards(width, height)
    add_r_pentomino(board_a)

    while (true) {
        step(board_a, board_b)
        var tmp = board_a;
        board_a = board_b;
        board_b = tmp;

        clear_screen()
        print_board(board_a);
        await sleep(100);
    }
}


if (require.main === module) {
    main();
}